### PR TITLE
feat(cli): expanded preview

### PR DESCRIPTION
**Pull request by louis-bompart**
## Proposed changes

When previewing, being it thru the dedicated command or part of the `push` flow allow the user to analyze files to preview what's what.

To be precise:
 - We 'pull' the target organization configuration (i.e. create a new snapshot from org, pull it).
 - We compare each file using the resourceKeys.
   - Resource key present in the target but not in the source snapshot are added
   - Resource key present in the source snapshot but not in the target are kept unless using the `-d` flag.
   - If present in both, then the value of the resource from the source is applied on the target.
   

## Testing

The focus of the tests will be on UT for this, testing is still ongoing:
   - [x] ExpandedPreviewer
   - [x] FileDiffProcessor

-----
CDX-347

### CODE REVIEW

File: packages/cli/src/lib/snapshot/expandedPreviewer.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
+import {
```
Comment (y-lakhdar) [inline packages/cli/src/lib/snapshot/expandedPreviewer.ts:1]: in that case, I would rename `reportViewer` to `Previewer`. That way, we will have the `Previewer` and the `ExpandedPreviewer`

File: packages/cli/tsconfig.json [hunk 2]
```diff
@@ -5,10 +5,11 @@
     "declaration": true,
     "importHelpers": true,
     "module": "commonjs",
+    "lib": ["ESNext"],
     "outDir": "lib",
     "rootDir": "src",
     "strict": true,
-    "target": "es2017",
+    "target": "es2019",
```
Comment (louis-bompart) [inline packages/cli/tsconfig.json:12]: Checked with node.green, '19 is ok for Node 14

File: packages/cli/src/lib/snapshot/expandedPreviewer.ts (hunk 1 above)
Comment (louis-bompart) [reply packages/cli/src/lib/snapshot/expandedPreviewer.ts:1]: Naming is really temporary, I'm mostly at the 'throw pasta on the wall and see what stick" stage.

And honestly, I don't think I would merge it till I can test it for real

Comment (louis-bompart) [reply packages/cli/src/lib/snapshot/expandedPreviewer.ts:1]: So @y-lakhdar, we go with `Previewer`?

**Review by olamothe — COMMENTED** [review]
@louis-bompart : Are these changes ready to review and merge ?

I saw your last comment about the backend not being ready and you preferring to wait. If that's the case, can you close this temporarily and re-open when it's ready (or switch it to draft).

Thanks !

**Comment by louis-bompart** [comment]
> @louis-bompart : Are these changes ready to review and merge ?

@olamothe, the PR description is up-to-date with the work that I think needs to be done (mainly, UT), but it is ready for review.

> I saw your last comment about the backend not being ready and you preferring to wait. If that's the case, can you close this temporarily and re-open when it's ready (or switch it to draft).

I don't know which comment you are talking about 😕?

**Comment by olamothe** [comment]
> I don't know which comment you are talking about 😕?

This one: 
```
And honestly, I don't think I would merge it till I can test it for real. (the more I coded it, the more I realized that there are a lot of places where it can go haywire, so I don't feel comfortable going forward till the backend is done and that I'm able to see what my code does.
```

But if it's ready, then 👍 Ill do the review.

File: packages/cli/src/lib/snapshot/expandedPreviewer.ts [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+import dedent from 'ts-dedent';
+import {Dirent} from 'fs';
+
+type ResourcesJSON = Object & {resourceName: string};
+
+type SnapshotFileJSON = Object & {
+  resources: Partial<{[key in ResourceSnapshotType]: ResourcesJSON[]}>;
+};
+
+export class ExpandedPreviewer {
```
[the remark this answers is not in the document]
Comment (louis-bompart) [reply packages/cli/src/lib/snapshot/expandedPreviewer.ts:21]: Agree. We have several loose ends on what is the 'coveo project' and how to manage it.

I'd say we could leave it as-is with a TODO, wdyt?

[the remark this answers is not in the document]
Comment (olamothe) [reply packages/cli/src/lib/snapshot/expandedPreviewer.ts:21]: I've created a PR https://github.com/coveo/cli/pull/379 which covers the topic a bit.

[the remark this answers is not in the document]
Comment (louis-bompart) [reply packages/cli/src/lib/snapshot/expandedPreviewer.ts:21]: Awesome thanks a lot 🙌.

#379 looks simple enough to merge.
Once it is, I will tweak out the code of ExpandedPreviewer to uses it and address this separation of concerns issue.

- [ ] Merge #379
- [ ] Refactor `.coveo/preview` check/creation to uses what's done in #379

File: packages/cli/src/lib/project/project.ts [hunk 4]
```diff
@@ -6,18 +6,18 @@ import {InvalidProjectError} from '../errors';
 import extract from 'extract-zip';
 
 export class Project {
-  public constructor(private pathToProject: string) {}
+  public constructor(private _pathToProject: string) {}
```
[the remark this answers is not in the document]
Comment (louis-bompart) [reply packages/cli/src/lib/project/project.ts:9]: The idea with the `public get` is that the property can be read outside the class but not modified.
So, the getter is necessary I think to reflect that.

File: packages/cli/src/lib/project/project.ts [hunk 5]
```diff
@@ -47,17 +46,21 @@ export class Project {
...
       cli.error(error);
     }
   }
 
-  private get pathToTemporaryZip() {
+  public get pathToProject() {
+    return this._pathToProject;
+  }
+
+  public get temporaryZipPath() {
```
[the remark this answers is not in the document]
Comment (louis-bompart) [reply packages/cli/src/lib/project/project.ts:59]: publicly readable, not modifiable, so I think not.

File: packages/cli/src/lib/snapshot/expandedPreviewer.ts [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+
+        const currentFile = readJSONSync(currentFilePath);
+
+        const diffedJSON = this.buildDiffedJson(currentFile, previewFile);
+
+        writeJSONSync(join(currentDir, file.name), diffedJSON, {
+          spaces: 2,
+        });
+      }
+    });
```
Comment (y-lakhdar) [inline packages/cli/src/lib/snapshot/expandedPreviewer.ts:134]: What happens if a file exist in `currentDir` but not in `previewDir`?

Comment (louis-bompart) [reply packages/cli/src/lib/snapshot/expandedPreviewer.ts:134]: For now, nothing.

🔜:tm: 
 - if shouldDelete is true, and the file ain't in previewDir, then we delete the file.
 - if shouldDelete is false, do nothing.

Comment (louis-bompart) [reply packages/cli/src/lib/snapshot/expandedPreviewer.ts:134]: - [ ] handle deleted files with shouldDelete flag

**Comment by louis-bompart** [comment]
Ready for full review. Incorporate part of #391 and needs it to be merged first
 - [ ] #391 merged

**Review by y-lakhdar — CHANGES_REQUESTED** [review]
image

There is point in the terminal where nothing happens while the expanded preview is being computed. Maybe a spinner or any kind of feedback would let the user know this is normal

**Review by y-lakhdar — APPROVED** [review]
GG.
Maybe another to add a spinner while computing the expanded preview

**Pull request merged**